### PR TITLE
Speed up linalg.dot

### PR DESCRIPTION
**Pull request by dstrain115**
The current linalg.combinators.dot uses np.linalg_multi_dot which does some fancy calculations to figure out the correct order to apply the dot product.  Since, most of the time, we are just multiplying matrices of the same size (2x2 or 4x4), there is no optimal order and we should just dot them in order.  This new function just does this and gets rid of the multi_dot call.  

I have kept the old functionality in case anyone needs it.

Here is the new output from https://github.com/quantumlib/Cirq/issues/2326 (shows over 100x improvement):
n: 0 time: 0.00013518333435058594
n: 100 time: 0.004878044128417969
n: 200 time: 0.00897073745727539
n: 300 time: 0.013186454772949219
n: 400 time: 0.018898725509643555
n: 500 time: 0.022220134735107422
n: 600 time: 0.02526688575744629
n: 700 time: 0.029251575469970703
n: 800 time: 0.034186601638793945
n: 900 time: 0.03860616683959961

### CODE REVIEW

File: cirq/linalg/combinators.py [hunk 1]
```diff
@@ -112,6 +114,31 @@ def dot(*values: Union[float, complex, np.ndarray]
...
+    sizes.  If you are using this for matrices of large and differing sizes,
+    consider using multi_dot for better performance
+
+    Args:
+        *values: The values to combine with the dot/matrix product.
+
+    Returns:
+        The resulting value or matrix.
+    """
+
```
Comment (dabacon) [inline cirq/linalg/combinators.py]: does this work for the len 0 case?

File: cirq/linalg/combinators.py [hunk 2]
```diff
@@ -112,6 +114,31 @@ def dot(*values: Union[float, complex, np.ndarray]
...
+    if len(values) == 1:
+        if isinstance(values[0], np.ndarray):
+            return np.array(values[0])
+        return values[0]
```
Comment (cduck) [inline cirq/linalg/combinators.py]: I think you meant `if not isintance...`.

You can juse use `return np.asarray(values[0])` here.

Comment (Strilanc) [reply cirq/linalg/combinators.py]: I wouldn't do that. The caller may assume that the output will not alias with the input.

File: cirq/linalg/combinators.py [hunk 3]
```diff
@@ -112,6 +114,31 @@ def dot(*values: Union[float, complex, np.ndarray]
     return np.linalg.multi_dot(values)
 
 
+def dot(*values: Union[float, complex, np.ndarray]
```
Comment (Strilanc) [inline cirq/linalg/combinators.py]: While we're tweaking this method we should add a `dtype=None` parameter for forcing the output dtype.

File: cirq/linalg/combinators.py (hunk 1 above)
Comment (Strilanc) [reply cirq/linalg/combinators.py]: I think we should purposefully fail in the 0 length case, to be safe. There's no way to know the user's expected shape. We could return the scalar 1, but that will almost definitely cause code to break in corner cases anyways.

Comment (dstrain115) [reply cirq/linalg/combinators.py]: It's not clear to me what you would want to return with no arguments, so I will just add a raise ValueError.

File: cirq/linalg/combinators.py (hunk 2 above)
Comment (dstrain115) [reply cirq/linalg/combinators.py]: I don't think this is correct.  I think the original intent was to return a copy of the array.

Comment (cduck) [reply cirq/linalg/combinators.py]: Then why `return values[0]`?

File: cirq/linalg/combinators.py (hunk 1 above)
Comment (viathor) [reply cirq/linalg/combinators.py]: The value of any inner product (and thus also dot product as a special case) is well-defined in the zero-dimensional space. This space consists of a single vector - the zero vector. And one of the inner product axioms explicitly demands that <0, 0> = 0.

This is also consistent with another widely used convention that calls for the neutral element to be the value of a zero-length sum/product/etc (so zero-length sum is zero, zero-length product is one etc). Dot product is defined as a sum, so empty dot is zero.

Comment (viathor) [reply cirq/linalg/combinators.py]: Numpy agrees:

```
>>> z = np.empty(shape=(0,))
>>> z
array([], dtype=float64)
>>> np.dot(z, z)
0.0
>>> 
```

Comment (dstrain115) [reply cirq/linalg/combinators.py]: I am not sure that is the correct analogy.  This is not a dot product in a zero-dimensional space.  Your example is a dot product of 2 elements (len=2) of zero-dimensions.  This is a dot product of no elements.  That doesn't make any sense and is likely a programming error.

As a counter point, numpy fails with the analogous case:

```
In [3]: np.linalg.multi_dot()                                                                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-aaa16d36e400> in <module>
----> 1 np.linalg.multi_dot()

TypeError: multi_dot() missing 1 required positional argument: 'arrays'
```

Comment (viathor) [reply cirq/linalg/combinators.py]: You're right. Sorry for not having read the full context.

Not even the rule about the neutral element can be applied here because without knowing the shape we wouldn't be able to pick the correct identity matrix.

File: cirq/linalg/combinators.py (hunk 2 above)
Comment (dstrain115) [reply cirq/linalg/combinators.py]: According to the mypy type annotation, if it's not an numpy array, it's a float or complex which are immutable.

File: cirq/linalg/__init__.py [hunk 4]
```diff
@@ -18,13 +18,8 @@
...
 
-from cirq.linalg.combinators import (
-    block_diag,
-    CONTROL_TAG,
-    dot,
-    kron,
-    kron_with_controls,
-)
+from cirq.linalg.combinators import (block_diag, CONTROL_TAG, dot, kron,
+                                     kron_with_controls, multi_dot)
```
Comment (dabacon) [inline cirq/linalg/__init__.py]: I think format made you do this because you don't have a trailing `,` after `multi_dot`?

File: cirq/linalg/combinators.py [hunk 5]
```diff
@@ -93,25 +93,56 @@ def kron_with_controls(*factors: Union[np.ndarray, complex, float]
...
+       ) -> Union[float, complex, np.ndarray]:
+    """Computes the dot/matrix product of a sequence of values.
+
+    Performs the computation in serial order without regard to the matrix
+    sizes.  If you are using this for matrices of large and differing sizes,
+    consider using multi_dot for better performance
```
Comment (dabacon) [inline cirq/linalg/combinators.py]: nit: missing period.

File: cirq/linalg/combinators.py (hunk 3 above)
Comment (dabacon) [reply cirq/linalg/combinators.py]: Do we have a use case for this?  `astype()` is an easy way to do this, I'd say wait until you see that this is commonly needed?

Comment (Strilanc) [reply cirq/linalg/combinators.py]: k

File: cirq/linalg/combinators.py [hunk 6]
```diff
@@ -93,25 +93,56 @@ def kron_with_controls(*factors: Union[np.ndarray, complex, float]
     return product
 
 
-def dot(*values: Union[float, complex, np.ndarray]
-        ) -> Union[float, complex, np.ndarray]:
+def multi_dot(*values: Union[float, complex, np.ndarray]
```
Comment (Strilanc) [inline cirq/linalg/combinators.py]: Why did you rename the method? This is a breaking change.

Comment (dstrain115) [reply cirq/linalg/combinators.py]: I am replacing this with another method.  I change the old one to multi_dot() and then creating a new dot() to replace it.  

Do you think I should just delete the old one instead?

Comment (Strilanc) [reply cirq/linalg/combinators.py]: I think we should just have `dot` and have it just skip all the optimization stuff. If people want that they can call numpy directly.

Comment (dstrain115) [reply cirq/linalg/combinators.py]: Ok, I deleted the old version.  Now it's just the serial dot product.
